### PR TITLE
[SYCL-MLIR] Allow argument promotion to perform on linkonce_odr functions

### PR DESCRIPTION
**Pull request by whitneywhtsang**
According to https://llvm.org/docs/LangRef.html#linkage-types:
> linkonce
Globals with “linkonce” linkage are merged with other globals of the same name when linkage occurs. This can be used to implement some forms of inline functions, templates, or other code which must be generated in each translation unit that uses it, but where the body may be overridden with a more definitive definition later. Unreferenced linkonce globals are allowed to be discarded. Note that linkonce linkage does not actually allow the optimizer to inline the body of this function into callers because it doesn’t know if this definition of the function is the definitive definition within the program or whether it will be overridden by a stronger definition. To enable inlining and other optimizations, use “linkonce_odr” linkage.
linkonce_odr
Some languages allow differing globals to be merged, such as two functions with different semantics. Other languages, such as C++, ensure that only equivalent globals are ever merged (the “one definition rule” — “ODR”). Such languages can use the linkonce_odr and weak_odr linkage types to indicate that the global will only be merged with equivalent globals. These linkage types are otherwise the same as their non-odr versions.

As argument promotion modifies all references in a translation unit, so it is safe to operate on functions with `linkonce_odr` linkage. To avoid conflicting with globals defined in other translation units, we modify the linkage to `internal` after the transformation.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp [hunk 1]
```diff
@@ -52,11 +52,28 @@ using namespace mlir;
...
-  auto sym = dyn_cast<SymbolOpInterface>(callableOp.getOperation());
-  return (sym && sym.isPrivate() && callableOp.getCallableRegion());
+/// Returns true if the callable operation \p callableOp has linkonce_odr
+/// linkage, and false otherwise.
+static constexpr StringRef linkageAttrName = "llvm.linkage";
+static bool isLinkonceODR(CallableOpInterface callableOp) {
+  if (!callableOp->hasAttr(linkageAttrName))
+    return false;
+  auto attr =
+      callableOp->getAttr(linkageAttrName).dyn_cast<mlir::LLVM::LinkageAttr>();
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp:62]: dyn_cast need to be a free function.

File: polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp [hunk 2]
```diff
@@ -52,11 +52,28 @@ using namespace mlir;
...
+  assert(attr && "Expecting LLVM::LinkageAttr");
+  return attr.getLinkage() == LLVM::Linkage::LinkonceODR;
+}
+
+// Change the linkage of \p callableOp from linkonce_odr to internel.
+// There can be globals of the same name (with linkonce_odr linkage) in another
+// translation unit. As they have different arguments, we need to change the
+// linkage of the modified function to internel.
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp:70]: typo: internal

File: polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp [hunk 3]
```diff
@@ -325,6 +342,7 @@ void Candidate::modifyCallee() {
 
   funcOp.setType(newFuncType);
   funcOp.setAllArgAttrs(newArgAttrs);
+  privatize(funcOp);
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp:345]: if (!isLinkonceODR(cast<CallableOpInterface>(callableOp.getOperation())))
   privatize(funcOp);

File: polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp [hunk 4]
```diff
@@ -52,11 +52,28 @@ using namespace mlir;
...
+  return attr.getLinkage() == LLVM::Linkage::LinkonceODR;
+}
+
+// Change the linkage of \p callableOp from linkonce_odr to internel.
+// There can be globals of the same name (with linkonce_odr linkage) in another
+// translation unit. As they have different arguments, we need to change the
+// linkage of the modified function to internel.
+static void privatize(FunctionOpInterface callableOp) {
+  if (!isLinkonceODR(cast<CallableOpInterface>(callableOp.getOperation())))
+    return;
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp:73]: remove, put in the caller and maybe assert here.

**Comment by whitneywhtsang** [comment]
The CI failures are fixed by https://github.com/intel/llvm/pull/8898.

**Review by etiotto — APPROVED** [review]
LGTM - thanks!

**Review by victor-eds — COMMENTED** [review]
What are the side-effects of privatizing and changing a function this way? Might this cause issues if the function is also defined in a different translation unit? I saw other passes modifying function signatures, e.g., [`-deadargelim`](https://llvm.org/docs/Passes.html#deadargelim-dead-argument-elimination) and [`-argpromotion`](https://llvm.org/docs/Passes.html#argpromotion-promote-by-reference-arguments-to-scalars) don't modify non-local functions. Shouldn't we do the same?

**Comment by whitneywhtsang** [comment]
> What are the side-effects of privatizing and changing a function this way? Might this cause issues if the function is also defined in a different translation unit? I saw other passes modifying function signatures, e.g., [`-deadargelim`](https://llvm.org/docs/Passes.html#deadargelim-dead-argument-elimination) and [`-argpromotion`](https://llvm.org/docs/Passes.html#argpromotion-promote-by-reference-arguments-to-scalars) don't modify non-local functions. Shouldn't we do the same?

Actually we are privatizing to prevent issues if the function is also defined in a different translation unit. If there exists in another translation unit, then the definition must be the same, due to `odr`, so to avoid different arguments, we need to privatize it. 

In LLVM, functions with `linkonce_odr` can be inlined, and removed when there are no more references in that translation unit. 
Our idea here is similar, we privatize it, as like cloning it, and removing the originally function.

**Comment by etiotto** [comment]
> > What are the side-effects of privatizing and changing a function this way? Might this cause issues if the function is also defined in a different translation unit? I saw other passes modifying function signatures, e.g., [`-deadargelim`](https://llvm.org/docs/Passes.html#deadargelim-dead-argument-elimination) and [`-argpromotion`](https://llvm.org/docs/Passes.html#argpromotion-promote-by-reference-arguments-to-scalars) don't modify non-local functions. Shouldn't we do the same?
> 
> Actually we are privatizing to prevent issues if the function is also defined in a different translation unit. If there exists in another translation unit, then the definition must be the same, due to `odr`, so to avoid different arguments, we need to privatize it.
> 
> In LLVM, functions with `linkonce_odr` can be inlined, and removed when there are no more references in that translation unit. Our idea here is similar, we privatize it, as like cloning it, and removing the originally function.

Functions with `linkonce_odr` can be merged by the linker. That means all definitions in the entire program must be the same. Therefore privatizing a definition in a particular translation unit would not lead to unresolved references at link time because other translation units must have their own weak definition. Note that this is different from functions that have `linkonce` linkage because that type of linkage does not require all definitions to be mergeable.

**Comment by victor-eds** [comment]
> Functions with `linkonce_odr` can be merged by the linker. That means all definitions in the entire program must be the same. Therefore privatizing a definition in a particular translation unit would not lead to unresolved references at link time because other translation units must have their own weak definition. Note that this is different from functions that have `linkonce` linkage because that type of linkage does not require all definitions to be mergeable.

Okey, I see. LGTM in that case. Thanks for clarifying!

**Pull request merged**